kafka-client/src/main/java/dev/responsive/kafka/api/stores/TtlProvider.java

```diff
@@ -116,56 +150,78 @@ private enum TtlType {
   private final TtlType ttlType;
   private final TtlDuration defaultTtl;
 
-  // Only 1 of these is used per instance, others will be null
-  private final Function<K, Optional<TtlDuration>> ttlForKey;
-  private final Function<V, Optional<TtlDuration>> ttlForValue;
-  private final BiFunction<K, V, Optional<TtlDuration>> ttlForKeyAndValue;
+  // Only non-null for key/value-based ttl providers
+  private final Serde<K> keySerde;
+  private final Serde<V> valueSerde;
+
+  private final BiFunction<K, V, Optional<TtlDuration>> computeTtl;
 
-  public TtlProvider(
+  private TtlProvider(
       final TtlType ttlType,
       final TtlDuration defaultTtl,
-      final Function<K, Optional<TtlDuration>> ttlForKey,
-      final Function<V, Optional<TtlDuration>> ttlForValue,
-      final BiFunction<K, V, Optional<TtlDuration>> ttlForKeyAndValue
+      final BiFunction<K, V, Optional<TtlDuration>> computeTtl,
+      final Serde<K> keySerde,
+      final Serde<V> valueSerde
   ) {
     this.ttlType = ttlType;
     this.defaultTtl = defaultTtl;
-    this.ttlForKey = ttlForKey;
-    this.ttlForValue = ttlForValue;
-    this.ttlForKeyAndValue = ttlForKeyAndValue;
+    this.keySerde = keySerde;
+    this.valueSerde = valueSerde;
+    this.computeTtl = computeTtl;
+  }
+
+  public Serde<K> keySerde() {
+    return keySerde;
+  }
+
+  public Serde<V> valueSerde() {
+    return valueSerde;
   }
 
   public TtlDuration defaultTtl() {
     return defaultTtl;
   }
 
-  public TtlDuration computeTtl(
+  public boolean hasTtl() {
```

Comment:
nit: this method name is a little confusing, you could have "no ttl" with a key-provider if the key provider returns INFINITE, right? perhaps we should invert usage and name it `hasNoTtl` (which is guaranteed if the condition is flipped)

Reply:
Agreed, I actually removed this method entirely in my last update